### PR TITLE
8309129: AArch64: guarantee(T != T2S) failed: "incorrect arrangement" after JDK-8307795

### PR DESCRIPTION
**Pull request by changpeng1997**
This patch fixes the issue introduced by JDK-8307795. Since addv[1] cannot support "T2S" SIMD arrangement, we should use addp[2] in this case.

TEST passed on AArch64:
hotspot:compiler/vectorapi, jdk:jdk/incubator/vector, tier1-3

[1]: https://developer.arm.com/documentation/ddi0596/2020-12/SIMD-FP-Instructions/ADDV--Add-across-Vector-
[2]: https://developer.arm.com/documentation/ddi0596/2020-12/SIMD-FP-Instructions/ADDP--vector---Add-Pairwise--vector--

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8309129](https://bugs.openjdk.org/browse/JDK-8309129): AArch64: guarantee(T != T2S) failed: "incorrect arrangement" after JDK-8307795

### Reviewers
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**) ⚠️ Review applies to [a5342348](https://git.openjdk.org/jdk/pull/14245/files/a5342348d1d6abd0feb66bb25bbab36ee7631cf5)
 * [Xiaohong Gong](https://openjdk.org/census#xgong) (@XiaohongGong - Committer)
 * [Evgeny Astigeevich](https://openjdk.org/census#eastigeevich) (@eastig - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/14245/head:pull/14245` \
`$ git checkout pull/14245`

Update a local copy of the PR: \
`$ git checkout pull/14245` \
`$ git pull https://git.openjdk.org/jdk.git pull/14245/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 14245`

View PR using the GUI difftool: \
`$ git pr show -t 14245`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/14245.diff">https://git.openjdk.org/jdk/pull/14245.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/14245#issuecomment-1569934418)

### CODE REVIEW

**Comment by changpeng1997** [comment]
@TobiHartmann Hi, could you please help to test and review this patch? Thanks!

**Comment by TobiHartmann** [comment]
Sure, I submitted testing and will report back once it passed.

File: src/hotspot/cpu/aarch64/aarch64_vector.ad [hunk 1]
```diff
@@ -5522,12 +5522,12 @@ instruct vstoremask_truecount_neon(iRegINoSp dst, vReg src, immI_gt_1 size, vReg
     // Input "src" is a vector mask represented as lanes with
     // 0/-1 as element values.
     uint esize = (uint)$size$$constant;
-    if (esize == 8) {
-      __ addpd($vtmp$$FloatRegister, $src$$FloatRegister);
+    uint length_in_bytes = Matcher::vector_length_in_bytes(this, $src);
+    Assembler::SIMD_Arrangement arrangement = Assembler::esize2arrangement(esize,
+                                                                           /* isQ */ length_in_bytes == 16);
+    if (arrangement == __ T2D || arrangement == __ T2S) {
```
Comment (eastig) [inline src/hotspot/cpu/aarch64/aarch64_vector.ad:5528]: I see `jdk/incubator/vector/Float64VectorTests.java` covers the case `arrangement == __ T2S`.
Is there a test covering the case `arrangement == __ T2D`?

**Comment by azvegint** [comment]
Please note that the associated test is now in the problem list, see #14250

**Comment by dcubed-ojdk** [comment]
jdk/incubator/vector/Float64VectorTests.java has been ProblemListed.
Please update this PR and remove the ProblemListing update before
integrating this fix.

**Comment by changpeng1997** [comment]
@azvegint @dcubed-ojdk Thanks, I will remove it.

File: src/hotspot/cpu/aarch64/aarch64_vector.ad (hunk 1 above)
Comment (changpeng1997) [reply src/hotspot/cpu/aarch64/aarch64_vector.ad:5528]: > I see `jdk/incubator/vector/Float64VectorTests.java` covers the case `arrangement == __ T2S`. Is there a test covering the case `arrangement == __ T2D`?

Thanks for your review.
I don't find such a case. I will add a test case in TestVectorMaskTrueCount.java to cover `arrangement == __ T2D`.

**Comment by TobiHartmann** [comment]
Testing of v00 passed. I'll re-run testing once you updated the PR.

File: test/hotspot/jtreg/compiler/vectorapi/TestVectorMaskTrueCount.java [hunk 2]
```diff
@@ -58,11 +59,11 @@ public class TestVectorMaskTrueCount {
         }
     }
 
-    static int maskAndTrueCountFloat(boolean[] a, boolean[] b, int idx) {
+    static int maskAndTrueCount(boolean[] a, boolean[] b, int idx, int SPECIES_length) {
```
Comment (eastig) [inline test/hotspot/jtreg/compiler/vectorapi/TestVectorMaskTrueCount.java:62]: `SPECIES_length` confuses me. The first confusing thing is `SPECIES` in capital letters. The second is the name itself. From the code I see it has meaning of `count`. Maybe using `count` instead would be better?

**Review by TobiHartmann — APPROVED** [review]
Looks good to me. Testing passed.

**Comment by changpeng1997** [comment]
> Looks good to me. Testing passed.

Thanks!

File: test/hotspot/jtreg/compiler/vectorapi/TestVectorMaskTrueCount.java (hunk 2 above)
Comment (changpeng1997) [reply test/hotspot/jtreg/compiler/vectorapi/TestVectorMaskTrueCount.java]: Thanks, I have modified the name of this parameter.

**Review by XiaohongGong — APPROVED** [review]
LGTM!

**Comment by changpeng1997** [comment]
@eastig Hi, Could you please help to review this patch ? Thanks!

**Review by eastig — APPROVED** [review]
LGTM

**Pull request closed without merging**